ext/dom/html_document.c

```diff
@@ -0,0 +1,1047 @@
+{
+	if (!decoding_encoding_ctx->fast_path) {
+		/* Fast path handles codepoints one by one, so this part is not applicable in that case */
+		(void) lxb_encoding_decode_finish(&decoding_encoding_ctx->decode);
+		size_t decoding_buffer_size = lxb_encoding_decode_buf_used(&decoding_encoding_ctx->decode);
+		if (decoding_buffer_size > 0) {
+			const lxb_codepoint_t *codepoints_ref = (const lxb_codepoint_t *) decoding_encoding_ctx->codepoints;
+			const lxb_codepoint_t *codepoints_end = codepoints_ref + decoding_buffer_size;
+			(void) decoding_encoding_ctx->encode_data->encode(&decoding_encoding_ctx->encode, &codepoints_ref, codepoints_end);
+			if (!dom_process_parse_chunk(ctx, document, parser, lxb_encoding_encode_buf_used(&decoding_encoding_ctx->encode), decoding_encoding_ctx->encoding_output, decoding_buffer_size, tokenizer_error_offset, tree_error_offset)) {
+				return false;
+			}
+		}
+	}
+	(void) lxb_encoding_encode_finish(&decoding_encoding_ctx->encode);
+	if (lxb_encoding_encode_buf_used(&decoding_encoding_ctx->encode) && !dom_process_parse_chunk(ctx, document, parser, lxb_encoding_encode_buf_used(&decoding_encoding_ctx->encode), decoding_encoding_ctx->encoding_output, lxb_encoding_decode_buf_used(&decoding_encoding_ctx->decode), tokenizer_error_offset, tree_error_offset)) {
+		return false;
+	}
+	return true;
+}
+
+static bool check_options_validity(zend_long options)
+{
+	const zend_long VALID_OPTIONS = XML_PARSE_NOERROR | XML_PARSE_COMPACT | HTML_PARSE_NOIMPLIED | DOM_HTML_NO_DEFAULT_NS;
+	if ((options & ~VALID_OPTIONS) != 0) {
+		zend_argument_value_error(2, "contains invalid flags (allowed flags: LIBXML_NOERROR, LIBXML_COMPACT, LIBXML_HTML_NOIMPLIED, DOM\\NO_DEFAULT_NS)");
+		return false;
+	}
+	return true;
+}
+
+PHP_METHOD(DOM_HTMLDocument, createEmpty)
+{
+	const char *encoding = "UTF-8";
+	size_t encoding_len = strlen("UTF-8");
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "|s", &encoding, &encoding_len) == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	const lxb_encoding_data_t *encoding_data = lxb_encoding_data_by_name((const lxb_char_t *) encoding, encoding_len);
+
+	if (encoding_data == NULL) {
+		zend_argument_value_error(1, "must be a valid document encoding");
+		RETURN_THROWS();
+	}
+	
+#ifdef LIBXML_HTML_ENABLED
+	xmlDocPtr lxml_doc = htmlNewDocNoDtD(NULL, NULL);
+	if (UNEXPECTED(lxml_doc == NULL)) {
+		goto oom;
+	}
+#else
+	xmlDocPtr lxml_doc = xmlNewDoc((const xmlChar *) "1.0");
+	if (UNEXPECTED(lxml_doc == NULL)) {
+		goto oom;
+	}
+	lxml_doc->type = XML_HTML_DOCUMENT_NODE;
+#endif
+
+	lxml_doc->encoding = xmlStrdup((const xmlChar *) encoding);
+
+	dom_object *intern = php_dom_instantiate_object_helper(return_value, dom_html_document_class_entry, (xmlNodePtr) lxml_doc, NULL);
+	intern->document->is_modern_api_class = true;
+	return;
+
+oom:
+	php_dom_throw_error(INVALID_STATE_ERR, 1);
+	RETURN_THROWS();
```

Comment:
Does Lebxor allow to register custom allocators?

Reply:
Got my answer and yes, so that case shouldn't be able to be hit as PHP will bail out before, maybe add a test with a very low memory limit to see the behaviour?